sdk/storage/azure-storage-common/inc/azure/storage/common/paged_response.hpp

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <type_traits>
+
+#include <azure/core/context.hpp>
+#include <azure/core/http/raw_response.hpp>
+
+namespace Azure { namespace Storage {
+
+  namespace _internal {
+    template <class Derived> class PageResult {
+    public:
+      std::unique_ptr<Azure::Core::Http::RawResponse> RawResponse;
+      std::string NextPageToken;
+
+      bool HasMore() const { return !NextPageToken.empty(); }
+
+      void NextPage()
+      {
+        if (!HasMore())
+        {
+          // NextPage() should NEVER be called if HasMore() is false
+          std::abort(); // or throw std::out_of_range
+        }
+
+        static_assert(
+            std::is_base_of<PageResult, Derived>::value,
+            "The template argument \"Derived\" should derive from PageResult<Derived>.");
+
+        static_cast<Derived*>(this)->OnNextPage(m_context);
+      }
+
+    protected:
+      PageResult() = default;
+      ~PageResult() = default;
+      PageResult(PageResult&&) = default;
+      PageResult& operator=(PageResult&&) = default;
+
+      Azure::Core::Context m_context;
+    };
+
+  } // namespace _internal
+
+  template <class T> class Pageable {
```

Comment:
Question:

Should we have this `Pageable` at all?
We can move `Iterator` to `PageResult<T>`.

But I think, `Pageable` works like a collection, so naturally it's iterable. `PageResult<T>` corresponds to one page, itself is not iterable, it may contain some fields that are iterable.